File: src/main/java/com/fish/extendedae_plus_client/util/UtilKeyBuilder.java
package com.fish.extendedae_plus_client.util;

import com.fish.extendedae_plus_client.ExtendedAEPlusClient;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.network.chat.contents.TranslatableContents;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.ItemLike;
import net.neoforged.neoforge.data.loading.DatagenModLoader;
import net.neoforged.neoforge.fluids.FluidType;
import net.neoforged.neoforge.registries.DeferredHolder;

import java.util.*;
import java.util.function.BiConsumer;

/// copy来的工具类就是好用😋
public class UtilKeyBuilder {
    public static final String creativeTab = "creative_tab.%s%s";
    public static final String tooltip = "tooltip.%s%s";
    public static final String screenTooltip = "tooltip.screen.%s%s";
    public static final String message = "message.%s%s";
    public static final String actionBar = "message.actionbar.%s%s";
    public static final String screen = "screen.%s%s";
    public static final String keywordGroup = "keywordGroup.%s%s";
    public static final String config = "%s.configuration%s";
    public static final String key = "key.%s%s";
    public static final String keyCategory = "key.category.%s%s";
    public static final String viewerInfo = "recipe_viewer.info.%s%s";
    public static final String viewerTooltip = "recipe_viewer.tooltip.%s%s";
    public static final String viewerCategory = "recipe_viewer.category.%s%s";
    public static final String jadeInfo = "jade.info.%s%s";
    public static final String jadeConfig = "config.jade.plugin_%s%s";

    // 高雅人士正在盗窃工具类.jpg

    public static BuilderGeneric<?> of(String keyTemplate) {
        var builder = new BuilderGeneric<>();
        builder.keyTemplate = keyTemplate;
        return builder;
    }

    public static BuilderGeneric<?> of(DeferredHolder<?, ?> holder) {
        var builder = new BuilderGeneric<>();
        builder.mainDescription = switch (holder.get()) {
            case Item item -> item.getDescriptionId().toLowerCase();
            case FluidType fluidType -> fluidType.getDescriptionId().toLowerCase();
            default -> ExtendedAEPlusClient.MODID;
        };
        return builder;
    }

    public static BuilderDataGen ofDataGen(String keyTemplate) {
        BuilderDataGen.checkEnvironment();
        var builder = new BuilderDataGen();
        builder.keyTemplate = keyTemplate;
        return builder;
    }

    public static BuilderDataGen ofDataGen(DeferredHolder<?, ?> holder) {
        BuilderDataGen.checkEnvironment();
        var builder = new BuilderDataGen();
        builder.mainDescription = switch (holder.get()) {
            case Item item -> item.getDescriptionId().toLowerCase();
            case FluidType fluidType -> fluidType.getDescriptionId().toLowerCase();
            default -> ExtendedAEPlusClient.MODID;
        };
        return builder;
    }

    public static class BuilderGeneric<TBuilder extends BuilderGeneric<TBuilder>> {
        protected String keyTemplate = "%s%s";
        protected String mainDescription = ExtendedAEPlusClient.MODID;
        protected String additionalKey = "";
        protected Object[] args;

        private BuilderGeneric() {
        }

        protected BuilderGeneric(BuilderGeneric<TBuilder> original) {
            this.copyFrom(original);
        }

        @SuppressWarnings("unchecked")
        private TBuilder self() {
            return (TBuilder) this;
        }

        public BuilderCollection bindCollection(Collection<Component> target) {
            return new BuilderCollection((BuilderCollection) this, target);
        }

        public BuilderCollection newArrayList() {
            return this.bindCollection(new ArrayList<>());
        }

        @SuppressWarnings("unchecked")
        public <TKey> BuilderMap<TKey> bindMap(Map<TKey, Component> target) {
            return new BuilderMap<>((BuilderGeneric<BuilderMap<TKey>>) this, target);
        }

        public BuilderMap<String> newHashMap() {
            return this.bindMap(new HashMap<>());
        }

        protected void copyFrom(BuilderGeneric<TBuilder> original) {
            this.keyTemplate = original.keyTemplate;
            this.mainDescription = original.mainDescription;
            this.additionalKey = original.additionalKey;
            this.args = original.args;
        }

        public TBuilder type(String keyTemplate) {
            this.keyTemplate = keyTemplate;
            return this.self();
        }

        public TBuilder item(ItemLike item) {
            this.mainDescription = item.asItem().getDescriptionId().toLowerCase();
            return this.self();
        }

        public TBuilder item(ItemStack itemStack) {
            return this.item(itemStack.getItem());
        }

        public TBuilder addStr(String additionalKey) {
            if (!this.additionalKey.endsWith("."))
                this.additionalKey += ".";
            this.additionalKey += additionalKey;
            return this.self();
        }

        public TBuilder addStr(boolean condition, String additionalKey) {
            if (condition) this.addStr(additionalKey);
            return this.self();
        }

        public TBuilder addStr(boolean condition, String keyA, String keyB) {
            return this.addStr(condition ? keyA : keyB);
        }

        public TBuilder args(Object... args) {
            this.args = Arrays.stream(args)
                    .map(object -> {
                        if (object == null) return "";
                        else if (!TranslatableContents.isAllowedPrimitiveArgument(object))
                            return object.toString();
                        else return object;
                    }).toArray();
            return this.self();
        }

        public TBuilder args(boolean condition, Object... args) {
            if (condition) this.args(args);
            return this.self();
        }

        public String buildRaw() {
            // For test
            // HelperI18nKeySaver.recordKey(currentKey);
            return String.format(this.keyTemplate, this.mainDescription, this.additionalKey);
        }

        public MutableComponent build() {
            var currentKey = this.buildRaw();

            if (this.args == null) return Component.translatable(currentKey);
            else return Component.translatable(currentKey, this.args);
        }
    }

    public static class BuilderDataGen extends BuilderGeneric<BuilderDataGen> {
        private static final Map<String, BiConsumer<String, String>> translators = new HashMap<>();
        private static final ThreadLocal<String> selectedLocale = ThreadLocal.withInitial(() -> "en_us");

        private BuilderDataGen() {
        }

        protected BuilderDataGen(BuilderDataGen original) {
            super(original);
        }

        public BuilderDataGen branch(String additionalKey, String value, String locale) {
            new BuilderDataGen(this)
                    .addStr(additionalKey)
                    .buildInto(value, locale);
            return this;
        }

        public BuilderDataGen branch(String additionalKey, String value) {
            return this.branch(additionalKey, value, selectedLocale.get());
        }

        public void buildInto(String value, String locale) {
            translators.getOrDefault(locale, (k, v) -> {}).accept(this.buildRaw(), value);
        }

        public void buildInto(String value) {
            this.buildInto(value, selectedLocale.get());
        }

        public static void bindTranslator(String locale, BiConsumer<String, String> translator) {
            checkEnvironment();
            translators.put(locale, translator);
            selectedLocale.set(locale);
        }

        public static void destroy(String locale) {
            checkEnvironment();
            translators.remove(locale);
            selectedLocale.remove();
        }

        public static void checkEnvironment() {
            if (!DatagenModLoader.isRunningDataGen())
                throw new IllegalStateException("Cannot use data-only methods outside of the runData phase");
        }
    }

    public abstract static class BuilderSnapshotable<TBuilder extends BuilderSnapshotable<TBuilder>>
            extends BuilderGeneric<TBuilder> {
        protected TBuilder snapshot = null;

        protected BuilderSnapshotable(BuilderGeneric<TBuilder> original, boolean saveSnapshot) {
            super(original);
            if (saveSnapshot) this.saveSnapshot();
        }

        public abstract TBuilder saveSnapshot();

        protected void restoreSnapshot() {
            if (this.snapshot == null) return;
            this.copyFrom(this.snapshot);
        }
    }

    public static class BuilderCollection extends BuilderSnapshotable<BuilderCollection> {
        private final Collection<Component> target;

        protected BuilderCollection(BuilderGeneric<BuilderCollection> original, Collection<Component> target) {
            this(original, target, true);
        }

        protected BuilderCollection(BuilderGeneric<BuilderCollection> original,
                                    Collection<Component> target,
                                    boolean saveSnapshot) {
            super(original, saveSnapshot);
            this.target = target;
        }

        @Override
        public BuilderCollection saveSnapshot() {
            this.snapshot = new BuilderCollection(this, this.target, false);
            return this;
        }

        public BuilderCollection buildInto() {
            this.target.add(this.build());
            this.restoreSnapshot();
            return this;
        }

        public BuilderCollection buildInto(String additionalKey) {
            this.target.add(this.addStr(additionalKey).build());
            this.restoreSnapshot();
            return this;
        }

        public Collection<Component> getCollection() {
            return this.target;
        }
    }

    public static class BuilderMap<TKey> extends BuilderSnapshotable<BuilderMap<TKey>> {
        private final Map<TKey, Component> target;

        protected BuilderMap(BuilderGeneric<BuilderMap<TKey>> original, Map<TKey, Component> target) {
            this(original, target, true);
        }

        protected BuilderMap(BuilderGeneric<BuilderMap<TKey>> original,
                             Map<TKey, Component> target,
                             boolean saveSnapshot) {
            super(original, saveSnapshot);
            this.target = target;
        }

        @Override
        public BuilderMap<TKey> saveSnapshot() {
            this.snapshot = new BuilderMap<>(this, this.target, false);
            return this;
        }

        /// 不建议真的在非string的情况下调用🤓
        public BuilderMap<TKey> buildInto(TKey key) {
            this.target.put(key, this.addStr(key.toString()).build());
            this.restoreSnapshot();
            return this;
        }

        public BuilderMap<TKey> buildIntoPlain(TKey key) {
            this.target.put(key, this.build());
            this.restoreSnapshot();
            return this;
        }

        public Map<TKey, Component> getMap() {
            return this.target;
        }
    }
}

File: src/main/kotlin/com/fish/extendedae_plus_client/util/UtilKeyBuilder.kt
package com.fish.extendedae_plus_client.util

import com.fish.extendedae_plus_client.ExtendedAEPlusClient
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.MutableComponent
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.ItemLike
import net.minecraftforge.data.loading.DatagenModLoader
import net.minecraftforge.fluids.FluidType
import net.minecraftforge.registries.RegistryObject
import java.util.function.BiConsumer

@Suppress("ConstPropertyName")
object UtilKeyBuilder {
    const val tooltip = "tooltip.%s%s"
    const val screenTooltip = "tooltip.screen.%s%s"
    const val message = "message.%s%s"
    const val actionBar = "message.actionbar.%s%s"
    const val screen = "screen.%s%s"
    const val keywordGroup = "keywordGroup.%s%s"
    const val config = "%s.configuration%s"
    const val key = "key.%s%s"
    const val keyCategory = "key.category.%s%s"
    const val viewerInfo = "recipe_viewer.info.%s%s"
    const val viewerTooltip = "recipe_viewer.tooltip.%s%s"
    const val viewerCategory = "recipe_viewer.category.%s%s"
    
    // 高雅人士正在重写工具类.jpg

    @JvmStatic
    fun of(keyTemplate: String): BuilderGeneric<*> {
        return BuilderGeneric().template(keyTemplate)
    }

    @JvmStatic
    fun of(holder: RegistryObject<*>): BuilderGeneric<*> {
        return BuilderGeneric().item(holder.get())
    }

    @JvmStatic
    fun ofDataGen(keyTemplate: String): BuilderDataGen {
        BuilderDataGen.checkEnvironment()
        return BuilderDataGen().template(keyTemplate)
    }

    @JvmStatic
    fun ofDataGen(holder: RegistryObject<*>): BuilderDataGen {
        BuilderDataGen.checkEnvironment()
        return BuilderDataGen().item(holder.get())
    }

    open class BuilderGeneric<TBuilder : BuilderGeneric<TBuilder>>(val builder: BuilderGeneric<TBuilder>?) {
        internal var keyTemplate = "%s%s"
        internal var mainDescription = ExtendedAEPlusClient.MODID
        internal var additionalKey = ""
        internal var args: Array<out Any> = arrayOf()
        
        internal constructor() : this(null)

        init {
            this.builder?.let(::copyFrom)
        }

        @Suppress("UNCHECKED_CAST")
        private fun unwrap() = this as TBuilder

        internal fun copyFrom(builder: BuilderGeneric<TBuilder>) {
            this.keyTemplate = builder.keyTemplate
            this.mainDescription = builder.mainDescription
            this.additionalKey = builder.additionalKey
            this.args = builder.args
        }
        
        fun bindCollection(target: MutableCollection<Component>) =
            BuilderCollection(this as BuilderCollection, target)
        
        fun newArrayList() = this.bindCollection(ArrayList())
        
        @Suppress("UNCHECKED_CAST")
        fun <TKey> bindMap(target: MutableMap<TKey, Component>) =
            BuilderMap(this as BuilderGeneric<BuilderMap<TKey>>, target)
        
        fun <TKey> newHashMap() = this.bindMap<TKey>(HashMap())

        fun template(template: String): TBuilder {
            this.keyTemplate = template
            return this.unwrap()
        }

        fun item(obj: Any?): TBuilder {
            this.mainDescription = when (obj) {
                is RegistryObject<*> -> return this.item(obj.get())
                is ItemStack -> return this.item(obj.item)
                is ItemLike -> obj.asItem().descriptionId.lowercase()
                is FluidType -> obj.descriptionId.lowercase()
                else -> ExtendedAEPlusClient.MODID
            }
            return this.unwrap()
        }

        fun addStr(additionalKey: String): TBuilder {
            if (!this.additionalKey.endsWith("."))
                this.additionalKey += "."
            this.additionalKey += additionalKey
            return this.unwrap()
        }

        fun addStr(predicate: Boolean, additionalKey: String): TBuilder {
            if (predicate) this.addStr(additionalKey)
            return this.unwrap()
        }

        fun addStr(predicate: Boolean, keyA: String, keyB: String) =
            this.addStr(if (predicate) keyA else keyB)

        fun args(vararg args: Any): TBuilder {
            this.args = args
            return this.unwrap()
        }

        fun buildRaw() =
            String.format(this.keyTemplate, this.mainDescription, this.additionalKey)

        fun build(): MutableComponent =
            Component.translatable(this.buildRaw(), *this.args)
    }

    class BuilderDataGen internal constructor(original: BuilderDataGen?) : BuilderGeneric<BuilderDataGen>(original) {
        internal constructor() : this(null)
        
        fun branch(additionalKey: String, value: String, locale: String): BuilderDataGen {
            BuilderDataGen(this)
                .addStr(additionalKey)
                .buildInto(value, locale)
            return this
        }

        fun branch(additionalKey: String, value: String) =
            this.branch(additionalKey, value, locale.get())

        fun buildInto(value: String, locale: String) {
            translators[locale]?.accept(this.buildRaw(), value)
        }

        fun buildInto(value: String) = this.buildInto(value, locale.get())

        companion object {
            private val translators: MutableMap<String, BiConsumer<String, String>> = HashMap()
            private val locale: ThreadLocal<String> = ThreadLocal.withInitial { "en_us" }

            fun bindTranslator(locale: String, translator: BiConsumer<String, String>) {
                this.checkEnvironment()
                this.locale.set(locale)
                this.translators[locale] = translator
            }

            fun destroy(locale: String) {
                this.checkEnvironment()
                this.locale.remove()
                this.translators.remove(locale)
            }

            internal fun checkEnvironment() {
                check(DatagenModLoader.isRunningDataGen())
                { "Cannot use data-only methods outside of the runData phase" }
            }
        }
    }

    abstract class BuilderSnapshotable<TBuilder : BuilderSnapshotable<TBuilder>> : BuilderGeneric<TBuilder> {
        internal var snapshot: TBuilder? = null
        
        internal constructor(original: BuilderGeneric<TBuilder>?, saveSnapshot: Boolean) : super(original) {
            if (saveSnapshot)
                this.snapshot()
        }
        
        abstract fun snapshot(): TBuilder
        
        internal fun restore() {
            this.snapshot?.let(::copyFrom)
        }
    }
    
    class BuilderCollection internal constructor(
        original: BuilderGeneric<BuilderCollection>?,
        private val target: MutableCollection<Component>,
        saveSnapshot: Boolean
    ) : BuilderSnapshotable<BuilderCollection>(original, saveSnapshot) {
        internal constructor(
            original: BuilderGeneric<BuilderCollection>?,
            target: MutableCollection<Component>
        ) : this(original, target, true)

        override fun snapshot(): BuilderCollection {
            this.snapshot = BuilderCollection(this, this.target, false)
            return this
        }

        fun buildInto(): BuilderCollection {
            this.target.add(this.build())
            this.restore()
            return this
        }

        fun buildInto(additionalKey: String): BuilderCollection {
            this.target.add(this.addStr(additionalKey).build())
            this.restore()
            return this
        }

        fun get(): Collection<Component> = this.target
    }
    
    class BuilderMap<TKey> internal constructor(
        original: BuilderGeneric<BuilderMap<TKey>>?,
        private val target: MutableMap<TKey, Component>,
        saveSnapshot: Boolean
    ) : BuilderSnapshotable<BuilderMap<TKey>>(original, saveSnapshot) {
        internal constructor(
            original: BuilderGeneric<BuilderMap<TKey>>?,
            target: MutableMap<TKey, Component>
        ) : this(original, target, true)

        override fun snapshot(): BuilderMap<TKey> {
            this.snapshot = BuilderMap(this, this.target, false)
            return this
        }

        /** 不建议真的在非string的情况下调用🤓 */
        fun buildInto(key: TKey): BuilderMap<TKey> {
            this.target[key] = this.addStr(key.toString()).build()
            this.restore()
            return this
        }

        fun buildIntoPlain(key: TKey): BuilderMap<TKey> {
            this.target[key] = this.build()
            this.restore()
            return this
        }
        
        fun get(): Map<TKey, Component> = this.target
    }
}